CoreCLR runtime tests on Mono Windows x64.

lateralusX opened the pull request:
Analyzed all failures around CoreCLR runtime tests on Mono Windows x64.

Several of the failing test suites (started with 44 failing test suites), are due to missing features, not implemented on Mono on any platform, but many of these tests are not run on none Windows platforms, so have not been an issue in the past. Explicitly included these tests in issues.targets for all mono runtime platforms where we don't run the test suite on none Windows platforms and fail on Windows due to unsupported features in Mono runtime.

Analyzed several test suites that can be fixed in Mono runtime to pass on all platforms, adding fixes to this PR and enable the tests:

- Several fixes to native custom marshaller.
- Fixes to extended error reporting in dl_* functions, needed by several native library loader tests.
- Fix for marshaling issues related to System.Drawings tests.
- Excluded parts of runtime tests on Mono depending on Windows SEH (similar to what's run on none Windows platforms).
- Port of https://github.com/mono/mono/pull/21365
- Fixed failures in valloc on Windows causing assert in OOM situations preventing tests to correctly pass.

PR adds Mono Windows x64 runtime tests to default CI runs.

lateralusX committed:
Fix native custom marshaler logic, passing ICustomMarshaler runtime test.

lateralusX committed:
Disable runtime tests unsupported/not working on Mono Windows.

Commit:
Enable windows libraries tests and runtime tests for mono

lambdageek commented on src/mono/mono/metadata/cominterop.c:
```diff
@@ -2104,7 +2104,7 @@ cominterop_get_ccw_method (MonoClass *iface, MonoMethod *method, MonoError *erro
 	m.mb = mb;
 	m.runtime_marshalling_enabled = TRUE;
-	mono_marshal_emit_managed_wrapper (mb, sig_adjusted, mspecs, &m, adjust_method, 0);
+	mono_marshal_emit_managed_wrapper (mb, sig_adjusted, mspecs, &m, adjust_method, 0, NULL);
```
@lateralusX are we going to build on Windows-like platforms with COM support enabled? You may want to also pick up https://github.com/mono/mono/pull/21365 which fixes a coop handle leak which leads to incorrect refcounts on some CCW objects

lateralusX replied:
For now I will keep it disabled, but maybe we should include that PR anyways? Do you want me to include it in this PR?

lateralusX replied:
I fixed the use of mono_marshal_emit_managed_wrapper to always require a MonoError.

lambdageek replied:
> For now I will keep it disabled, but maybe we should include that PR anyways? Do you want me to include it in this PR?

Yea, I think let's pick up those changes.

Long-term I'm not sure if we should re-enable cominterop.c in dotnet/runtime.  I think we're better off implementing ComWrappers.  But I suppose that depends on the use-cases.

lateralusX replied:
> Long-term I'm not sure if we should re-enable cominterop.c in dotnet/runtime. I think we're better off implementing ComWrappers. But I suppose that depends on the use-cases.

Jupp, we will need need to fix unmanaged member function support for delegates in Mono to implement ComWrappers.

lateralusX replied:
Backport of https://github.com/mono/mono/pull/21365, https://github.com/dotnet/runtime/pull/64281/commits/38fa8ad7577d3fa05b1526173d9dbb293d7625e2

lambdageek commented on src/mono/mono/metadata/marshal-ilgen.c:
```diff
@@ -6193,8 +6267,47 @@ emit_marshal_variant_ilgen (EmitMarshalContext *m, int argnum, MonoType *t,
+static gboolean
+emit_managed_wrapper_validate_signature (MonoMethodSignature* sig, MonoMarshalSpec** mspecs, MonoError* error)
+{
+	if (mspecs && error) {
```
I don't like using `MonoError` as a flag like this. Let's keep `MonoError` usage predictable and boring in the entire runtime.

lambdageek commented on src/mono/mono/metadata/native-library.c:
```diff
@@ -1219,9 +1219,11 @@ ves_icall_System_Runtime_InteropServices_NativeLibrary_FreeLib (gpointer lib, Mo
-		MonoDl raw_module = { { 0 } };
-		raw_module.handle = lib;
-		mono_dl_close (&raw_module);
+		MonoDl *raw_module = (MonoDl *) g_malloc0 (sizeof (MonoDl));
+		if (raw_module) {
+			raw_module->handle = lib;
+			mono_dl_close (raw_module);
+		}
```
why?

Also, leaks the newly-allocated memory.

lateralusX replied:
mono_dl_close does a g_free as part of its implementation, that's what previously caused a corruption when using it as a stack variable, so the new allocated memory won't leak, it will be freed in mono_dl_close, https://github.com/dotnet/runtime/blob/4da6b9a8d55913c0ea560d63590d35dc942425be/src/mono/mono/utils/mono-dl.c#L391

lateralusX committed:
Fix correct parsing of -mono and -excludemonofailures in tests/build.cmd.

lateralusX committed:
Review feedback.

lateralusX committed:
Arguments not recongnized by build.cmd must be appended last to arg list.

lambdageek commented on src/mono/mono/metadata/marshal-ilgen.c:
```diff
@@ -6193,8 +6267,47 @@ emit_marshal_variant_ilgen (EmitMarshalContext *m, int argnum, MonoType *t,
+				MonoMarshalType *marshal_type = mono_marshal_load_type_info (mono_class_from_mono_type_internal (sig->params [i]));
+				for (int field_idx = 0; field_idx < marshal_type->num_fields; ++field_idx) {
+					if (marshal_type->fields [field_idx].mspec && marshal_type->fields [field_idx].mspec->native == MONO_NATIVE_CUSTOM) {
+						mono_error_set_generic_error (error, "System", "TypeLoadException", "Value type includes custom marshaled fields");
```
```suggestion
						MonoClass *klass = mono_class_from_mono_type_internal (sig->params [i]);
						mono_error_set_type_load_class (error, klass, "Value type includes custom marshaled fields");
```

(could also hoist `mono_class_from_mono_type_internal` from line 6295 above)

lateralusX committed:
Exclude BestFitMapping interop tests, unsupported on Mono.

lateralusX committed:
Fix build error.

lateralusX committed:
Review feedback.

lateralusX committed:
Fix build.

lateralusX committed:
Fix crash in Drawings test on Windows Mono.

lateralusX committed:
Disable test to search for test causing OOM on Windows CI machines.

lateralusX committed:
Fix build error.

lateralusX committed:
Make sure we throw OOM on failure to alloc aligned memory on Windows.

lateralusX committed:
Fix ComDisabled test to handle exception throwed by Marshal.NoCom.cs.

lateralusX committed:
Disable test triggering OOM on CI.

lateralusX committed:
Disable MarshalComDisabledTests on Mono.

lateralusX committed:
Run Windows library tests in runtime-extra-platforms.

lateralusX committed:
Port of https://github.com/mono/mono/pull/21365.

lateralusX committed:
Restore GetTypeFromCLSID behavior in Marshal.NoCom.cs

lateralusX committed:
Fix whitespace.

lambdageek reviewed (approved):
I wonder if we the native probing difference will be an issue in setups with libraries for multiple platforms available

steveisok reviewed (approved):
Nice job!

lateralusX commented on src/mono/mono/metadata/native-library.c:
```diff
@@ -519,18 +522,27 @@ netcore_probe_for_module (MonoImage *image, const char *file_name, int flags)
+		mono_error_cleanup (error);
+		error_init_reuse (error);
+		module = netcore_probe_for_module_variations (pinvoke_search_directories[i], file_name, lflags, error);
+		if (!module && !is_ok (error) && mono_error_get_error_code (error) == MONO_ERROR_BAD_IMAGE)
```
Yes, had that in mind as well, but since only Windows will trigger a MONO_ERROR_BAD_IMAGE at the moment, I didn't pursue it further. I can take a quick look to see if we can postpone the error and if the test is still passing doing it that way.

lateralusX replied:
I made a change where we will keep any last MONO_ERROR_BAD_IMAGE error (using mono_error_move), but continue probing. If we exhaust all probing attempts but still haven't been able to load a module, and have encountered a MONO_ERROR_BAD_IMAGE failure in any of the probing attempts, we will raise that error.

I believe this will be inline how it worked before, we will exhaust all probing attempts and complete as soon as we successfully loaded a module found in the probing paths. If we exhaust all probing attempts, but have hit a MONO_ERROR_BAD_IMAGE in any of the attempts, we will raise that exception (inline with the expectations of src\tests\Interop\NativeLibrary\AssemblyLoadContext\ResolveUnmanagedDllTests.csproj), if not we will raise the regular DLL not found exception.

lambdageek replied:
Thanks!

And thanks for reminding me about `mono_error_move`.

lateralusX committed:
Review feedback.

lateralusX committed:
Fix incorrect free of stack memory.

# Conflicts:
#	src/mono/mono/metadata/native-library.c

lateralusX committed:
Fix NativeLibraryTests on Mono.

# Conflicts:
#	src/mono/mono/metadata/native-library.c

lateralusX committed:
Update issues.targets with issues.

# Conflicts:
#	src/tests/Interop/NativeLibrary/Callback/CallbackStressTest_TargetWindows.csproj

lateralusX committed:
Disable a couple of new R2R tests on Mono Windows.

lateralusX committed:
Disable a couple of new MonoAPI tests on Mono Windows.
